### PR TITLE
Improvement/add real docs

### PR DESCRIPTION
**Pull request by slashformotion**
improve the documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,26 +1,102 @@
...
+- [Contributing](#contributing)
+- [Licence](#licence)
+
+# Quickstart
+
+You can either use the [Docker Install (WIP)](<!-- TODO: link the docker install instructions here and remove the WIP mention-->) or build it from source .
+
+## Docker install
+
+(WIP)
```
Comment (Zorin95670) [inline README.md:35]: nope, please complete

File: README.md [hunk 2]
```diff
@@ -1,26 +1,102 @@
...
+export BADAAS_DATABASE_HOST=<complete>
+export BADAAS_DATABASE_DBNAME=<complete>
+export BADAAS_DATABASE_SSLMODE=<complete>
+export BADAAS_DATABASE_USERNAME=<complete>
+export BADAAS_DATABASE_PASSWORD=<complete>
+./badaas 
 ```
 
-Once all is done, you have a binary `badaas` at the root of the project.
+### Configuration
```
Comment (Zorin95670) [inline README.md:71]: add section for each themes of configuration, list and explain all commands

File: README.md [hunk 3]
```diff
@@ -1,26 +1,102 @@
...
+
+The config file can be placed at `/etc/badaas/badaas.yml` or `$HOME/.config/badaas/badaas.yml` or in the same folder as the badaas binary `./badaas.yml`. 
+
+If needed, the location can be overridden using the config key `config_path`. 
 
-## Development
+# Tutorial
 
-### Directory structure
+(WIP)
```
Comment (Zorin95670) [inline README.md:95]: complete or remove

File: README.md [hunk 4]
```diff
@@ -1,26 +1,102 @@
...
+
+# Principes et architecture
+
+## Directory structure
```
Comment (Zorin95670) [inline README.md:99]: move in contributing

File: README.md [hunk 5]
```diff
@@ -33,52 +109,27 @@ badaas
 ├ resources            ⇨ Contains all the applications resources, like constants or other.
 ├ router               ⇨ Contains the route definitions for application.
 ├ scripts              ⇨ Contains shell scripts, the Docker files and docker-compose files for e2e test.
+├ validators           ⇨ Contains validator for various data "standards"
 ```
 
-### E2E testing
+## E2E testing
```
Comment (Zorin95670) [inline README.md:115]: move in contributing

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (slashformotion) [reply README.md:35]: ben on a pas d'image docker officielle donc avec xavier on s'était dit qu'on compléterai plus tard.

Comment (slashformotion) [reply README.md:35]: docker build -t imgname .

File: badaas.yml.example [hunk 6]
```diff
@@ -1,45 +1,55 @@
...
+  # (mandatory)
   password: postres
-  
-  # The settings for the initialization of the database
-  init:
 
-    # Number of time badaas will try to establish a connection to the database
+  # The settings for the initialization of the database 
+  init:
+    # Number of time badaas will try to establish a connection to the database.
```
Comment (Zorin95670) [inline badaas.yml.example:24]: missing default

File: badaas.yml.example [hunk 7]
```diff
@@ -1,45 +1,55 @@
...
     retry: 10
 
-    # Waiting time between connection
+    # Waiting time between connection, in seconds.
```
Comment (Zorin95670) [inline badaas.yml.example:27]: missing default

File: badaas.yml.example [hunk 8]
```diff
@@ -1,45 +1,55 @@
...
-
-  # The maximum timout for read/write ops.
-  timeout: 15 # in seconds
-
-  pagination:
-    page:
-    # The maximum number of record per page 
-      max: 50
-
-# The settings for the database.
```
Comment (Zorin95670) [inline badaas.yml.example:16]: keep explanation of section

File: badaas.yml.example [hunk 9]
```diff
@@ -1,45 +1,55 @@
...
+  # The settings for the initialization of the database 
+  init:
+    # Number of time badaas will try to establish a connection to the database.
     retry: 10
 
-    # Waiting time between connection
+    # Waiting time between connection, in seconds.
     retryTime: 5
 
+server:
```
Comment (Zorin95670) [inline badaas.yml.example:30]: add section explanation

File: badaas.yml.example [hunk 10]
```diff
@@ -1,45 +1,55 @@
...
+  # The address to bind badaas to.
+  # Defaults to "0.0.0.0"
+  host: "" 
+
+  # The port badaas should use.
+  # Defaults to 8000
+  port: 8000
+
+  # The maximum timout for read/write ops.
```
Comment (Zorin95670) [inline badaas.yml.example:39]: ops?

File: badaas.yml.example [hunk 11]
```diff
@@ -1,45 +1,55 @@
...
+  # Defaults to 15
+  timeout: 15 # in seconds
+
+  pagination:
```
Comment (Zorin95670) [inline badaas.yml.example:43]: missing section explanation

File: badaas.yml.example [hunk 12]
```diff
@@ -1,45 +1,55 @@
...
+    page:
+      # The maximum number of record per page 
+      # Defaults to 100
+      max: 100
+
 logger:
```
Comment (Zorin95670) [inline badaas.yml.example:49]: missing section explanation

File: badaas.yml.example [hunk 13]
```diff
@@ -1,45 +1,55 @@
...
+  # Either dev or prod
```
Comment (Zorin95670) [inline badaas.yml.example:50]: for all options value please use `this`

File: README.md [hunk 14]
```diff
@@ -1,84 +1,99 @@
...
+  - [Docker install](#docker-install)
+  - [Install from sources](#install-from-sources)
+    - [Prerequisites](#prerequisites)
+    - [Configuration](#configuration)
+- [Contributing](#contributing)
+- [Licence](#licence)
+
+# Quickstart
+
+You can either use the [Docker Install (WIP)](<!-- TODO: link the docker install instructions here and remove the WIP mention-->) or build it from source .
```
Comment (Zorin95670) [inline README.md:27]: remove wip

File: README.md [hunk 15]
```diff
@@ -1,84 +1,99 @@
...
+
+## Docker install
+
+You can build the image using `docker build -t badaas .` since we don't have an official docker image yet.
+
+## Install from sources
+
+### Prerequisites
+
+Get the sources of the project, either by visiting the [releases]() page and downloading an archive or clone the main branch (please be aware that is it not a stable version).
```
Comment (Zorin95670) [inline README.md:37]: rewrite this and put TODO to explain that we have to set release link note here

File: README.md [hunk 16]
```diff
@@ -1,84 +1,99 @@
...
+
+To build the project:
 
 - [Install go](https://go.dev/dl/#go1.18.4) v1.18
 - Install project dependencies
-```
+
+```txt
```
Comment (Zorin95670) [inline README.md:44]: txt -> bash

File: README.md [hunk 17]
```diff
@@ -1,84 +1,99 @@
...
-- Create a pull request on github for this branch into `main`.
-- Once the pull request validated and merged, tag the `main` branch with `vX.Y.Z`
-- After the tag is pushed, make the release on the tag in GitHub
+The config file can be placed at `/etc/badaas/badaas.yml` or `$HOME/.config/badaas/badaas.yml` or in the same folder as the badaas binary `./badaas.yml`.
 
-### Git: Default branch
+If needed, the location can be overridden using the config key `config_path`.
 
-The default branch is main. Direct commit on it is forbidden. The only way to update the application is through pull request.
+***For a full overview of the configuration keys: please head to  [configuration.md](./configuration.md).***
```
Comment (Zorin95670) [inline README.md:91]: [configuration documentation](./configuration.md)

File: README.md [hunk 18]
```diff
@@ -1,84 +1,99 @@
...
 
-Release tag are only done on the `main` branch.
+# Contributing
 
-### Git: Branch naming policy
+See [CONTRIBUTING.md](./CONTRIBUTING.md).
```
Comment (Zorin95670) [inline README.md:95]: See this section

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
+<h1> Contribute to the development of badaas</h1>
```
Comment (Zorin95670) [inline CONTRIBUTING.md:1]: remove h1 and use #

Comment (Zorin95670) [reply CONTRIBUTING.md:1]: Modify all next section by adding a #
`# Tests -> ## Tests`

File: configuration.md [hunk 20]
```diff
@@ -0,0 +1,96 @@
+<h1> Configuration </h1>
+
+To see a simple example of a working config file: head to [`badaas.yml.example`](../badaas.yml.example)
```
Comment (Zorin95670) [inline configuration.md:3]: simple -> complete

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: README.md (hunk 14 above)
Comment (Zorin95670) [reply README.md:27]: not remove

File: badaas.yml.example [hunk 21]
```diff
@@ -1,45 +1,61 @@
-server:
```
Comment (Zorin95670) [inline badaas.yml.example:1]: Rename this file to `badaas.example.yml` to keep syntax highlighting and improve readability.

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: configuration.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+  password: postres
+
+  # The settings for the initialization of the database 
+  init:
+    # Number of time badaas will try to establish a connection to the database.
+    # Defaults to 10
+    retry: 10
+
+    # Waiting time between connection, in seconds.
+    # Defaults to 5
```
Comment (k-y0u) [inline configuration.md:53]: default

File: configuration.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    retryTime: 5
+```
+
+Please note that the init section `init:` is not mandatory. Badaas is suited with a simple but effective retry mecanism that will retry `database.init.retry` time to establish a connection with the database. Badaas will wait `database.init.retryTime` seconds between each retry.
+
+## Logger
+
+Badaas use a structured logger that can output json logs in production and user adapted logs for debug using the `logger.mode` key.  
+
+Badaas offer the possibility to change the log message of the Middleware Logger but provides a sane default. It is formated using the Jinja syntax. The values available are `method`, `url` and `protocol`.
```
Comment (k-y0u) [inline configuration.md:63]: offers

File: configuration.md [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+```yml
+# The settings for the logger.
+logger:
+  # Either `dev` or `prod`
+  # Defaults to `prod`
```
Comment (k-y0u) [inline configuration.md:69]: default

File: configuration.md [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+  mode: prod
+  request:
+    # Change the log emitted when badaas receive a request on a valid endpoint.
```
Comment (k-y0u) [inline configuration.md:72]: receives

File: configuration.md [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+
+You can change the host Badaas will bind to, the port and the timeout in seconds.
+
+Additionaly you can change the number of elements returned by default for a paginated response.
+
+```yml
+# The settings for the http server.
+server:
+  # The address to bind badaas to.
+  # Defaults to "0.0.0.0"
```
Comment (k-y0u) [inline configuration.md:86]: Default

File: configuration.md [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+  host: "" 
+
+  # The port badaas should use.
+  # Defaults to 8000
```
Comment (k-y0u) [inline configuration.md:90]: default

File: configuration.md [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+  port: 8000
+
+  # The maximum timout for the http server.
+  # Defaults to 15
```
Comment (k-y0u) [inline configuration.md:94]: default

File: configuration.md [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+  timeout: 15 # in seconds
```
Comment (k-y0u) [inline configuration.md:95]: choose the good place for the unit

File: configuration.md [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+
+  # The settings for the pagination.
+  pagination:
+    page:
+      # The maximum number of record per page 
+      # Defaults to 100
```
Comment (k-y0u) [inline configuration.md:101]: default

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: badaas.yml.example [hunk 31]
```diff
@@ -1,45 +1,61 @@
...
-  # The password of the account on the database server
+  # The password of the account on the database server.
+  # (mandatory)
   password: postres
-  
-  # The settings for the initialization of the database
-  init:
 
-    # Number of time badaas will try to establish a connection to the database
+  # The settings for the initialization of the database
```
[the remark this answers is not in the document]
Comment (slashformotion) [reply badaas.yml.example]: d

File: README.md [hunk 32]
```diff
@@ -1,84 +1,99 @@
...
-To build the project: 
+- **Authentification**: Badaas can authentify users using it's internal authentification scheme or externally by using protocols such as OIDC, SAML, Oauth2, 
+- **Habilitation**: On a ressource access, Badaas will check if the user is authorized using a RBAC model.
+- **Distribution**: Badaas is built to run in clusters by default. Communications between nodes are TLS encrypted using [shoset](https://github.com/ditrit/shoset). 
+- **Persistence**: Applicative objects are persisted as well as user files. These ressources are shared accross the cluster to increase resiliency.
+- **Querying Ressources**: Ressources are accessible via a REST API.
+- **Posix complient**: Badaas strive towards being a good unix citizen and respecting commonly accepted norms. (see [Configuration](#configuration))
+- **Advanced logs management**: Badaas provide an interface to interact with the logs produced by the cluster. Logs are formated in json by default.
+
+To quickly get up and running, please head to the [miniblog tutorial](<!-- TODO: link the miniblog tutorial here -->)
```
[the remark this answers is not in the document]
Comment (slashformotion) [reply README.md:15]: Isn't "get up and running" a valid expression ? https://www.linguee.fr/anglais-francais/traduction/up+and+running.html

File: CONTRIBUTING.md [hunk 33]
```diff
@@ -0,0 +1,122 @@
...
+- `docs/`: Contains the documentation.
+- `features/`: Contains all the feature tests (or end to end tests).
+- `logger/` *(Go code)*: Contains the logger creation logic. Please don't call it from your own services and code, use the dependency injection system.
+- `persistance/` *(Go code)*: 
+  - `/gormdatabase/` *(Go code)*: Contains the logic to create a <https://gorm.io> database. Also contains a go package named `gormzap`: it is a compatibility layer between *gorm.io/gorm* and *github.com/uber-go/zap*.
+  - `/models/` *(Go code)*: Contains the models. (For a structure to me considered a valid model, it has to embed `models.BaseModel` and satisfy the `models.Tabler` interface. This interface returns the name of the sql table.)
+    - `/dto/` *(Go code)*: Contains the Data Transfert Objects. They are used mainly to decode json payloads.
+  - `/pagination/` *(Go code)*: Contains the pagination logic.
+  - `/repository/` *(Go code)*: Contains the repository interface and implementation. Use uint as ID when using gorm models.
+- `ressources/` *(Go code)*: Contains the ressources shared with the rest of the codebase (ex: API version).
```
Comment (k-y0u) [inline CONTRIBUTING.md:64]: `ressources/`->`resources/`

File: README.md [hunk 34]
```diff
@@ -1,84 +1,99 @@
...
 
-Backend and Distribution as a Service
+Badaas enables the effortless construction of ***distributed, resilient, highly available and secure applications by design***, while ensuring very simple deployment and management (NoOps). 
 
-## Build
+Badaas provides several key features:
 
-To build the project: 
+- **Authentification**: Badaas can authentify users using its internal authentification scheme or externally by using protocols such as OIDC, SAML, Oauth2...
+- **Habilitation**: On a ressource access, Badaas will check if the user is authorized using a RBAC model.
```
Comment (k-y0u) [inline README.md:8]: ressource -> resource

File: CONTRIBUTING.md (hunk 33 above)
Comment (k-y0u) [reply CONTRIBUTING.md:64]: Contains the ressources -> Contains the resources

File: README.md [hunk 35]
```diff
@@ -1,84 +1,99 @@
...
+- **Distribution**: Badaas is built to run in clusters by default. Communications between nodes are TLS encrypted using [shoset](https://github.com/ditrit/shoset). 
+- **Persistence**: Applicative objects are persisted as well as user files. Those resources are shared accross the clusters to increase resiliency.
+- **Querying Resources**: Ressources are accessible via a REST API.
```
Comment (k-y0u) [inline README.md:11]: Ressources -> Resources

File: README.md (hunk 32 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply README.md:15]: So it should be `get it/badass/theproject up and running`.

File: README.md [hunk 36]
```diff
@@ -1,84 +1,99 @@
...
 
-### E2E testing
+### Configuration
 
-We use [godog](https://github.com/cucumber/godog) to run all e2e tests.
+Badaas use [verdeter](https://github.com/ditrit/verdeter) to manage it's configuration. So Badaas is POSIX complient by default.
 
-To execute E2E tests :
+Badaas can be configured using environment variables, configuration files or CLI flags.
+CLI flags takes priority on the environment variables and the environment variables take priority on the content of the configuration file.
```
[the remark this answers is not in the document]
Comment (k-y0u) [reply README.md]: only one ?

File: configuration.md [hunk 37]
```diff
@@ -0,0 +1,103 @@
+# Configuration
+
+To see a complete example of a working config file: head to [`badaas.example.yml`](./badaas.example.yml)
+
+As said in the README:
+
+> Badaas can be configured using environment variables, configuration files or CLI flags.
+> CLI flags take priority on the environment variables and the environment variables take priority on the content of the configuration files.
```
Comment (k-y0u) [inline configuration.md:8]: `configuration files` -> several ?

**Review by k-y0u — CHANGES_REQUESTED** [review]

**Review by k-y0u — APPROVED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Review by k-y0u — APPROVED** [review]

**Pull request merged**